src/coreclr/src/System.Private.CoreLib/src/System/RuntimeType.CoreCLR.cs

```diff
@@ -3873,43 +3867,71 @@ private void CreateInstanceCheckThis()
+                    _pfnCtor = (delegate*<object?, void>)RuntimeMethodHandle.GetFunctionPointer(rmh);
+                }
+
+                Debug.Assert(_pfnCtor != null);
             }
 
-            private void Initialize()
+            private void Initialize(RuntimeType type)
             {
-                if (!_hCtorMethodHandle.IsNullHandle())
-                {
-                    _ctorAttributes = RuntimeMethodHandle.GetAttributes(_hCtorMethodHandle);
+                Debug.Assert(type != null);
 
-                    // The default ctor path is optimized for reference types only
-                    ConstructorInfo delegateCtorInfo = s_delegateCtorInfo ??= typeof(CtorDelegate).GetConstructor(new Type[] { typeof(object), typeof(IntPtr) })!;
+                // If we reached this point, we already know that the construction information
+                // can be cached, so all of the runtime reflection checks succeeded. We just
+                // need to special-case Nullable<T> (to return null).
+                //
+                // No synchronization is needed in this method since we have marked the _pfnNewobj
+                // field as volatile, and if there's multi-threaded access all threads will agree
+                // on the values to write anyway.
+                //
+                // !! IMPORTANT !!
+                // Don't assign the function pointer return value of GetNewobjHelperFnPtr directly
+                // to the backing field, as setting the field marks initialization as complete.
+                // Be sure to perform any last-minute checks *before* setting the backing field.
 
-                    // No synchronization needed here. In the worst case we create extra garbage
-                    _ctor = (CtorDelegate)delegateCtorInfo.Invoke(new object?[] { null, RuntimeMethodHandle.GetFunctionPointer(_hCtorMethodHandle) });
+                delegate*<MethodTable*, object?> pfnNewobj = RuntimeTypeHandle.GetNewobjHelperFnPtr(type, out _pMT, unwrapNullable: false, allowCom: true);
+                if (_pMT->IsNullable)
+                {
+                    pfnNewobj = &GetNull; // Activator.CreateInstance(typeof(Nullable<T>)) => null
                 }
-                _isFullyInitialized = true;
+
+                Debug.Assert(pfnNewobj != null);
+                _pfnNewobj = pfnNewobj; // setting this field marks the instance as fully initialized
             }
 
-            public void EnsureInitialized()
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public void EnsureInitialized(RuntimeType type)
             {
-                if (!_isFullyInitialized)
-                    Initialize();
+                if (_pfnNewobj == null) // use this field's value as the indicator of whether initialization is finished
+                {
+                    Initialize(type);
+                }
             }
+
+            private static object? GetNull(MethodTable* pMT) => null;
```

Comment:
Another weird linker issue here. I tried defining this as a local static function within the _Initialize_ method. However, the linker was removing it, even though there was very clearly a `ldftn GetNullLocalFn` instruction in the IL stream. This was causing the runtime to fail with a `MethodMissingException` at JIT time.

Making the method a proper static function instead of a static local function resolved the issue.

Reply:
Seems to be resolved now.